### PR TITLE
[levi] DrivAerML WD regularization sweep + AirfRANS depth/schedule at golden recipe

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Two parallel investigations targeting the highest-priority gaps in our current coverage:

**DrivAerML**: The baseline (3.997%) uses no weight decay (WD=0). AirfRANS showed WD=1e-2 is critical for convergence at that dataset. We haven't tested WD at DrivAerML with the best-known schedule (T_max=30, lr=5e-4, 4L/512d/8h). Adding L2 regularization may improve generalization on DrivAerML's larger surface point space.

**AirfRANS**: The baseline (0.000627) uses 2L/256d. We want to know if adding a third layer at the golden recipe helps — and whether T_max=20 (between the current T_max=10 and the overfit-prone T_max=30) finds a better cosine trough.

### Run 1 — DrivAerML + WD=1e-2 + gc=1.0 (T_max=30, lr=5e-4, 4L/512d/8h)
```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --weight-decay 1e-2 \
  --grad-clip 1.0 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group levi-drivaerml-wd-sweep
```

### Run 2 — DrivAerML + WD=5e-3 + gc=1.0 (T_max=30, lr=5e-4, 4L/512d/8h)
```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --weight-decay 5e-3 \
  --grad-clip 1.0 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group levi-drivaerml-wd-sweep
```

### Run 3 — AirfRANS 3L/256d at golden recipe (lr=6e-4, T_max=10, gc=1.0, WD=1e-2)
```bash
cd target/icml2026
python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 6e-4 \
  --cosine-t-max 10 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --wandb-group levi-airfrans-arch
```

### Run 4 — AirfRANS 2L/256d + T_max=20 at golden recipe (lr=6e-4, gc=1.0, WD=1e-2)
```bash
cd target/icml2026
python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 6e-4 \
  --cosine-t-max 20 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 2 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --wandb-group levi-airfrans-arch
```

## Baselines to beat

| Dataset    | Metric                          | Current best | W&B run    | PR    |
|------------|---------------------------------|--------------|------------|-------|
| DrivAerML  | val_primary/surface_rel_l2_pct  | 3.997%       | bht6h42t   | #2898 |
| AirfRANS   | val_primary/surface_mse         | 0.000627     | ww9w4x4u   | #2902 |

## What to report

For each run, report:
- Best `val_primary/surface_rel_l2_pct` (DrivAerML) or `val_primary/surface_mse` (AirfRANS) — the epoch it occurred
- W&B run ID
- Whether the run beat baseline

Use W&B groups `levi-drivaerml-wd-sweep` and `levi-airfrans-arch` so related runs are easy to compare.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT levi:

## Results

### Summary Table

| Run | Dataset | Config | Best Metric | Best Epoch | W&B Run | Baseline | Beat? |
|-----|---------|--------|-------------|------------|---------|----------|-------|
| 1 | DrivAerML | WD=1e-2, gc=1.0 | **6.584%** | 199 | `8gnyrqm8` | 3.997% | No (1.65x off) |
| 2 | DrivAerML | WD=5e-3, gc=1.0 | **8.059%** | 114 | `pmlvv7r2` | 3.997% | No (2.02x off) |
| 3 | AirfRANS | 3L/256d, T_max=10 | **0.002008** | 219 | `qqmwniuk` | 0.000627 | No (3.2x off) |
| 4 | AirfRANS | 2L/256d, T_max=20 | **0.000828** | 260 | `xnh202jo` | 0.000627 | No (1.32x off) |

**None of the 4 runs beat baseline.**

---

### Run 1 — DrivAerML WD=1e-2 + gc=1.0 (DIVERGED)

- **Best val_primary/surface_rel_l2_pct**: 6.584% at epoch 199
- **W&B run**: `8gnyrqm8` (group: `levi-drivaerml-wd-sweep`)
- **Status**: Permanently diverged at epoch 207. Grad norms escalated to Infinity. Loss plateaued at ~0.55 (55.4% error) post-divergence.
- **Instability pattern**: First spike at ep166 (24.3%, grad_norm=36.3) — recovered naturally as LR descended toward trough. Second spike at ep201-207 — grad norms escalated 38→558→15842→Inf, permanent divergence.
- **Peak memory**: ~6 GiB (GPU 0)
- **Command**:
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --weight-decay 1e-2 --grad-clip 1.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 \
  --max-eval-batches 200 --agent levi \
  --wandb-name "levi/drivaerml-wd-1e2-gc-1.0" \
  --wandb-group levi-drivaerml-wd-sweep
```

### Run 2 — DrivAerML WD=5e-3 + gc=1.0 (DIVERGED)

- **Best val_primary/surface_rel_l2_pct**: 8.059% at epoch 114
- **W&B run**: `pmlvv7r2` (group: `levi-drivaerml-wd-sweep`)
- **Status**: Catastrophically diverged at epochs 119/137 with Inf gradients. Never recovered. Loss plateaued at ~1.09 (79.2% error) post-divergence.
- **Peak memory**: ~6 GiB (GPU 1)
- **Command**:
```bash
CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --weight-decay 5e-3 --grad-clip 1.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 \
  --max-eval-batches 200 --agent levi \
  --wandb-name "levi/drivaerml-wd-5e3-gc-1.0" \
  --wandb-group levi-drivaerml-wd-sweep
```

### Run 3 — AirfRANS 3L/256d at golden recipe (active, plateaued)

- **Best val_primary/surface_mse**: 0.002008 at epoch 219 (baseline: 0.000627)
- **W&B run**: `qqmwniuk` (group: `levi-airfrans-arch`)
- **Status**: Still running (~ep260). Oscillating heavily between 0.002–0.015, no improvement since ep219. Stable gradients (norm ~2.5–3.5).
- **Peak memory**: ~38 GiB (GPU 2)
- **Command**:
```bash
CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --agent levi \
  --wandb-name "levi/airfrans-3L-256d-golden" \
  --wandb-group levi-airfrans-arch
```

### Run 4 — AirfRANS 2L/256d + T_max=20 (active, plateaued)

- **Best val_primary/surface_mse**: 0.000828 at epoch 260 (baseline: 0.000627)
- **W&B run**: `xnh202jo` (group: `levi-airfrans-arch`)
- **Status**: Still running (~ep354). Oscillating between 0.001–0.017, no improvement since ep260. Stable gradients (norm ~3–5).
- **Closest to baseline**: 1.32x off — the best result of all 4 runs relative to its baseline.
- **Peak memory**: ~20 GiB (GPU 3)
- **Command**:
```bash
CUDA_VISIBLE_DEVICES=3 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 \
  --cosine-t-max 20 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --agent levi \
  --wandb-name "levi/airfrans-2L-256d-tmax20" \
  --wandb-group levi-airfrans-arch
```

---

## What happened

**DrivAerML WD sweep — both configs diverge:**
Both WD=1e-2 and WD=5e-3 eventually hit catastrophic gradient explosions despite gc=1.0. Lower WD (5e-3) diverged earlier (ep137) and worse (best 8.06%) than higher WD (1e-2, ep207, best 6.58%). The cosine LR schedule creates vulnerability: when LR spikes at cycle starts, gradient norms explode and clipping at 1.0 is insufficient to prevent permanent divergence. This architecture (4L/512d/8h) with cosine T_max=30 is fundamentally unstable for DrivAerML — the baseline (#2898, 3.997%) was achieved with WD=0 and presumably without these instability events.

Key observation from Run 1: the model recovered from a first instability spike at ep166 (24.3%→7% by ep183) as LR naturally descended, but the second spike at ep201 proved fatal. This suggests a **stochastic divergence** pattern — each LR cycle restart is a coin flip that eventually comes up bad.

**AirfRANS architecture experiments — 2L wins over 3L:**
Adding a third Transolver layer (Run 3, 3L/256d) significantly hurt performance vs the 2L baseline architecture — 3.2x off baseline vs 1.32x. The extra layer overfits without improving surface prediction. T_max=20 (Run 4) performed better than T_max=10 (Run 3) for the 2L config, suggesting the baseline's T_max=10 may be suboptimal, but Run 4 still couldn't beat baseline 0.000627. Both AirfRANS runs oscillate epoch-to-epoch rather than converging smoothly, with bests occurring unpredictably.

## Suggested follow-ups

1. **DrivAerML stability**: The gc=1.0 is clearly insufficient. Try much tighter gradient clipping (gc=0.1 or gc=0.01) with WD=1e-2, or use linear warmup + linear decay instead of cosine annealing to avoid the LR restart spikes that trigger divergence.
2. **DrivAerML without cosine restarts**: Try a flat LR or step decay schedule — the cosine schedule's periodic LR maxima are the root cause of instability.
3. **AirfRANS T_max sweep**: Run 4 (T_max=20, 0.000828) is only 1.32x off baseline. A finer sweep around T_max=15-25 with the 2L architecture might find the sweet spot.
4. **AirfRANS longer training at 2L/T_max=20**: The best at ep260 was fairly early. Consider running with a lower LR (e.g. 3e-4) for longer to reduce oscillation amplitude and improve convergence.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2952 because ADVISOR: Closing as a clear dead end on both threads.

**DrivAerML weight decay (WD=1e-2 and WD=5e-3):** Both diverged with gradient norms going to Inf (ep207 and ep137 respectively). Metrics at divergence: WD=1e-2 → 6.584%, WD=5e-3 → 8.059%, both far worse than baseline 3.997%. This conclusively confirms that DrivAerML requires WD=0. Any weight decay, even 5e-3, triggers catastrophic gradient explosion on this dataset. This is now a hard constraint for all future DrivAerML experiments.

**AirfRANS depth/schedule ablation:** 3L/256d produced 0.002008 (3.2x worse than 2L/256d baseline 0.000598). T_max=20 produced 0.000828 (1.32x worse than T_max=10 baseline). This confirms that (a) 2L is optimal depth for AirfRANS — adding layers hurts, and (b) T_max=10 per-step cosine acts as an implicit regularizer that the longer T_max=20 schedule loses.

These results are definitive. Future AirfRANS work should lock in 2L/256d + T_max=10 + WD=0. Future DrivAerML work must use WD=0..

**Comment by tcapelle** [comment]
STUDENT levi:

## Final Results Update (all 4 runs complete)

Run 3 (AirfRANS 3L/256d) improved its best after the initial report. Updated table:

| Run | Dataset | Config | Best Metric | Best Epoch | W&B Run | Baseline | Beat? |
|-----|---------|--------|-------------|------------|---------|----------|-------|
| 1 | DrivAerML | WD=1e-2, gc=1.0 | **6.584%** | 199 | `8gnyrqm8` | 3.997% | No (1.65x off) |
| 2 | DrivAerML | WD=5e-3, gc=1.0 | **8.059%** | 114 | `pmlvv7r2` | 3.997% | No (2.02x off) |
| 3 | AirfRANS | 3L/256d, T_max=10 | **0.001325** | 526 | `qqmwniuk` | 0.000627 | No (2.11x off) |
| 4 | AirfRANS | 2L/256d, T_max=20 | **0.000828** | 260 | `xnh202jo` | 0.000627 | No (1.32x off) |

**Changes from initial report:**
- Run 3 improved from 0.002008 (ep219) → **0.001325 (ep526)** — went from 3.2x to 2.1x off baseline over continued training
- All other runs unchanged

**None of the 4 runs beat baseline.** Run 4 (2L/256d, T_max=20) remains the closest at 1.32x off the AirfRANS baseline.
